statecheck/expect_matching_values.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package statecheck
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	tfjson "github.com/hashicorp/terraform-json"
+
+	"github.com/hashicorp/terraform-plugin-testing/tfjsonpath"
+)
+
+// Resource State Check
+var _ StateCheck = expectMatchingValues{}
+
+type expectMatchingValues struct {
+	resourceAddressOne string
+	attributePathOne   tfjsonpath.Path
+	resourceAddressTwo string
+	attributePathTwo   tfjsonpath.Path
+}
+
+// CheckState implements the state check logic.
+func (e expectMatchingValues) CheckState(ctx context.Context, req CheckStateRequest, resp *CheckStateResponse) {
+	var resourceOne, resourceTwo *tfjson.StateResource
+
+	if req.State == nil {
+		resp.Error = fmt.Errorf("state is nil")
+
+		return
+	}
+
+	if req.State.Values == nil {
+		resp.Error = fmt.Errorf("state does not contain any state values")
+
+		return
+	}
+
+	if req.State.Values.RootModule == nil {
+		resp.Error = fmt.Errorf("state does not contain a root module")
+
+		return
+	}
+
+	for _, r := range req.State.Values.RootModule.Resources {
+		if e.resourceAddressOne == r.Address {
+			resourceOne = r
+
+			break
+		}
+	}
+
+	if resourceOne == nil {
+		resp.Error = fmt.Errorf("%s - Resource not found in state", e.resourceAddressOne)
+
+		return
+	}
+
+	resultOne, err := tfjsonpath.Traverse(resourceOne.AttributeValues, e.attributePathOne)
+
+	if err != nil {
+		resp.Error = err
+
+		return
+	}
+
+	for _, r := range req.State.Values.RootModule.Resources {
+		if e.resourceAddressTwo == r.Address {
+			resourceTwo = r
+
+			break
+		}
+	}
+
+	if resourceTwo == nil {
+		resp.Error = fmt.Errorf("%s - Resource not found in state", e.resourceAddressTwo)
+
+		return
+	}
+
+	resultTwo, err := tfjsonpath.Traverse(resourceTwo.AttributeValues, e.attributePathTwo)
+
+	if err != nil {
+		resp.Error = err
+
+		return
+	}
+
+	if !reflect.DeepEqual(resultOne, resultTwo) {
+		resp.Error = fmt.Errorf("values are not equal: %v != %v", resultOne, resultTwo)
+
+		return
+	}
+}
+
+// ExpectMatchingValues returns a state check that asserts that the specified attributes at the given resources
+// have a matching value.
+func ExpectMatchingValues(resourceAddressOne string, attributePathOne tfjsonpath.Path, resourceAddressTwo string, attributePathTwo tfjsonpath.Path) StateCheck {
+	return expectMatchingValues{
+		resourceAddressOne: resourceAddressOne,
+		attributePathOne:   attributePathOne,
+		resourceAddressTwo: resourceAddressTwo,
+		attributePathTwo:   attributePathTwo,
+	}
+}
```

Comment:
Similar naming concerns here as I'm a little uneasy about the mixed terminology here:

* Previously "match" meant string regular expression testing
* Previously this 1:1 checking was called "pair"
* Currently with knownvalue checks, they are called "exact"

I'd also consider this along with my prior comment about having an interface/generic check for these sorts of checks, because maybe having something like that would be:

```go
statecheck.ExpectKnownValuePair(
  "examplecloud_thing.test",
  tfjsonpath.New("some_computed_attr"),
  "examplecloud_widget.test",
  tfjsonpath.New("some_other_attr"),
  knownvalue.StringExactPair(),
)
```

And we could still introduce a "wrapper" check if we really wanted.

Reply:
I'm in agreement. Think we can drop both `ExpectContain` and `ExpectMatchingValues` in favour of using `ExpectKnownValuePair`.

In terms of exact value matching, I was wondering whether we could forgo type specific value matching and use something like the following:

```go
package knownvalue

import (
	"fmt"
	"reflect"
)

var _ CheckPair = exactPair{}

type exactPair struct{}

// CheckValue determines whether the passed values are equal.
func (v exactPair) CheckValue(a, b any) error {
	if !reflect.DeepEqual(a, b) {
		return fmt.Errorf("values are not equal: %v != %v", a, b)
	}

	return nil
}

// ExactPair returns a Check for asserting whether two values are equal.
func ExactPair() exactPair {
	return exactPair{}
}
```

Reply:
As discussed out-of-band, have deprecated the TestCheckResourceAttrPair, and TestCheckTypeSetElemAttrPair functions without replacement and linked to https://github.com/hashicorp/terraform-plugin-testing/issues/282 in the docs for now.